google/google-how-to/gce/set-hostname-cloudinit.rst

```diff
@@ -0,0 +1,55 @@
+Set hostname with cloud-init on GCE
+===================================
+
+Hostnames can be set in many ways, including a user manually configuring the hostname, DHCP, and using a set-up tool like cloud-init.  On Ubuntu, most of these methods use the underlying ``hostname`` command, which sets the hostname statically if the ``/etc/hostname`` file exists and transiently if it does not.
+
+Historically, all Ubuntu images have an ``/etc/hostname`` file and cloud-init, when run, would create an ``/etc/hostname`` file if it did not already exist.  Starting in 24.04 GCE images, the ``/etc/hostname`` file is no longer present by default and the cloud-init key ``create_hostname_file`` set to false.  This means that when a user or tool uses ``hostname`` to set the hostname, it is set transiently and trying to set the instance hostname using solely user-supplied metadata to cloud-init or manually from the command line will be overwritten by Google's DHCP service.  It is possible to set up a GCE instance with a user defined hostname which ensures Google's DHCP-supplied hostname matches the desired hostname but if that is not done, the DHCP hostname will be a random string such as ``[INTANCE-ID].c.[PROJECT].internal``.  A common programmatic way to set hostnames for machines is to use cloud-init metadata through the gcloud flag ``--metadata-from-file user-data=[FILENAME]``, where ``[FILENAME]`` contains the cloud-init configurations desired on the VMs including the hostname.  Starting in 24.04 GCE images, only passing the ``hostname`` key to cloud-init will no longer statically set the instance hostname.
```

Comment:
1. This introduction/explanation is useful. Even though it seems to be more appropriate for an 'Explanation' page, let's leave it in for the time being. (If we decide to write a separate explanation page about using hostnames sometime in the future, then we can move this content there and put a link to it here.) 

2. 
> It is possible to set up a GCE instance with a user defined hostname which ensures Google's DHCP-supplied hostname matches the desired hostname 

This statement seems to be slightly inaccurate/misleading (based on Google's description given in the `Create a VM instance with a custom hostname` link) for two reasons:
(i) the 'user defined hostname' has to be a fully qualified DNS name and cannot just be a name like 'test123'.
(ii) even if a desired hostname (in the correct DNS format) is specified, Google still creates its own internal DNS name which can also be used to connect to the VM. So Google's DHCP-supplied hostname does not match our desired hostname. But both of these names are available for use.

Reply:
You are right that Google only accepts FQDNs so a user couldn't set their hostname to `test123`; they would have to use something like `test123.test.com`.  I'm not sure how/where to specify that.

Google will still create it's DNS name and you can connect to the machine using both names.  Google DHCP will provide both to the instance, with preference given to the user-specified one.  So if the user runs `hostname` on their machine, they will see the hostname they set, rather than `instance-1.c.ubuntu-catred.internal`.  If you don't use the `--hostname` flag on instance creation, when you check your own hostname, you see the internal one.